src/transformStateWithClones.js

```diff
@@ -7,7 +7,67 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const properties = {
+    addProperties: 'addProperties',
+    clear: 'clear',
+    removeProperties: 'removeProperties',
+  };
+  const newObj = { ...state };
+  let obj = {};
+  const newArray = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case properties.addProperties:
+        assignNewDataToState(newObj, action);
+        obj = writeToArray(obj, newObj, newArray);
```

Comment:
the code could be simplified to:
```suggestion
        obj = { ...newObj, ...action.extraData };
```